### DISCUSSION TITLE
Concept `machine` and `instance`

### DISCUSSION THREAD

**Discussion by karajan1001** [treeverse/dvc#6890] [New Features & Ideas]
Follow https://github.com/iterative/dvc/pull/6855#issuecomment-954984810

We have `machine`, `instance`, and config, status with them, They are a bit confusing in the current unreleased documents.
So maybe it's time to fix the name and usage for each of the concepts.

Currently, `machine` is a template or a Metaclass object, while `instance` represents the actual running server, generated from one `machine`.  So `machine` here is only an abstract concept, it is only a name of config and does not represent any Cloud/Server in the real world. With this `machine` config, we can generate an actual instance from it that is called `instance`. `instance` here actually represent a real word machine / virtual machine. It can't be configured directly. It only has status and has a similar value to the `machine` that generated it.

Currently, we have `dvc machine list` and `dvc machine status` these two similar commands, one is for `machine` to show the config from them, the other is for `instance` to show the running status. 

| name   | attributes of it | show cmd |
|:---:|:---:|:---:|
| machine | config | machine list |
| instance | status | machine status |

`machine` sometimes represents the concept of a config, sometimes represents a remote machine and it is also used in the first level in commands makes it more confusing. 

Another question is that should the `machine` and `instance` here be a one-to-one relationship?

@dberenbaum , @jorgeorpinel , @pmrowla

**Comment by dberenbaum** [comment]
Your table looks good to me as long as we are consistent with it. It seems to parallel docker images vs containers.

> Another question is that should the `machine` and `instance` here be a one-to-one relationship?

I don't think we should assume a 1-1 relationship, although for now only 1-1 is possible. Having a cluster of identical workers will definitely be needed. This could be handled either by having multiple instances or by having a single instance that defines a kubernetes cluster or something similar, but I don't see much downside in keeping open the option of having multiple instances per machine config.

**Reply by jorgeorpinel** [reply]
IMO the table doesn't clarify much. It's explained from the PoV of the developers. Let's think about the user perspective instead 🙂 

But what are we trying to decide on exactly? Whether to keep both concepts "machine" and "instance"? What commands to have?

Assuming we're keeping `dvc machine` as the command and feature name (I believe it has been discussed a lot already):

- `machine add` and `machine create` sound like the same to me, so maybe just keep one of those (`add` for consistency) for registering machine configurations.
  Or maybe `machine register` so it doesn't sound like an actual box is being built.
  Or `machine config:add` which is super explicit, but then there's a 3rd level of subcommands.
- `machine run` to (instantiate if needed) and start the instance.
  Maybe there's a separate `machine build` or `instantiate` command or `machine run --build-only` to do the first step explicitly.
  
  But maybe I'm misunderstanding the meaning of "running" (status). **That's another confusing concept here.** Do you need to start/stop machines so that experiments can be run on them? Or are they assumed to always be "on" and available to pick up experiment jobs? And does "running" mean it's available? Or that it's running an experiment at the moment?

- `machine list` can show everything: the machine (config) names, all of its instances, and the status of each one. It can accept a `name` argument to limit listing only instances of a specific machine, or even an `--instance` option to show only one specific known instance.
  `status` may not be great since we use it for `git` and `dvc status` with a very different meaning (related to changes in the project).

**Reply by dberenbaum** [reply]
> But what are we trying to decide on exactly? Whether to keep both concepts "machine" and "instance"? What commands to have?

I would say both. As far as concepts, I would propose keeping "machine" as the umbrella term for this feature. Following TPI, the "resource" could be the config entry and the "instance" could be the compute provider server.

>     * `machine add` and `machine create` sound like the same to me, so maybe just keep one of those (`add` for consistency) for registering machine configurations.
>       Or maybe `machine register` so it doesn't sound like an actual box is being built.
>       Or `machine config:add` which is super explicit, but then there's a 3rd level of subcommands.

I agree these sound too much alike and it's hard to remember which is which. `machine add` was used for consistency with `remote add` I think.

>     * `machine run` to (instantiate if needed) and start the instance.
>       Maybe there's a separate `machine build` or `instantiate` command or `machine run --build-only` to do the first step explicitly.
>       But maybe I'm misunderstanding the meaning of "running" (status). **That's another confusing concept here.** Do you need to start/stop machines so that experiments can be run on them? Or are they assumed to always be "on" and available to pick up experiment jobs? And does "running" mean it's available? Or that it's running an experiment at the moment?

`machine build` or maybe `machine start` make sense to me and are more explicit than `machine create`.

Regarding what "running" means, so far all of these commands do nothing but configure and build cloud compute instances. There is no actual work being done on them yet. The exact syntax for how work will be performed on those instances will be discussed as we start working on that part, but so far most suggestions have been via `dvc exp run --machine` or similar, so it might not reside under the `dvc machine` namespace at all.

As far as whether users need to stop/start machines, I think we should leave it flexible. The initial proposal assumed the instance would start automatically if it didn't already exist, and it could be destroyed at the end with a `--destroy` option, but this is not final. Users should not have to manually start/stop to do `dvc exp run --machine`, but it should probably be an option because there are some workflows where it can be useful:
* Adding more experiments ad hoc (without necessarily queueing them all at once) without having to rebuild each time.
* Connecting to always on instances like internal servers.
* Manually configuring or debugging something before running experiments (for instance, this can be one way to address how to get credentials onto the instance in the short term).

>     * `machine list` can show everything: the machine (config) names, all of its instances, and the status of each one. It can accept a `name` argument to limit listing only instances of a specific machine, or even an `--instance` option to show only one specific known instance.
>       `status` may not be great since we use it for `git` and `dvc status` with a very different meaning (related to changes in the project).

Since we seem to be heading in the direction of duplicating all the info between them, using a single command might be a reasonable approach. I would probably say `status` is more like an instance-level field in `dvc machine list`, where values could be something like "building", "idle", "running exp-123456", and "destroying."

**Reply by jorgeorpinel** [reply]
> `machine add` was used for consistency with remote add I think
`machine build` or maybe machine start make sense to me
using a single command might be a reasonable approach

Thanks @dberenbaum. I vote for `machine add`, `machine build` ("start" sounds like running too), and `machine list` then.

> so far most suggestions have been via dvc exp run --machine or similar, so it might not reside under the dvc machine namespace at all

Sounds good. So let's not use "run/running/runner" with `dvc machine` for now, I think.

**Reply by dberenbaum** [reply]
@karajan1001 @pmrowla What do you think?

**Reply by pmrowla** [reply]
I don't have any strong opinions about naming here, I'd be fine with `machine build` or `start`.

The distinction between `machine` and `instance` (and similarly the difference with `list` and `status`) is just that we have two different concepts we need to represent to the user here:

1. A common configuration for a remote `machine` type (i.e. cloud, CPU size/count, gpu size/type, storage size, etc). This is currently configured via the `config`/`remote` equivalent commands - `dvc machine list/add/remove`
2. One or more started/built/created instances of a `machine` configuration. This is currently configured via the `machine create/destroy/status` commands.

Regarding `list`/`status`, we could have a single command that shows both the user's machine configurations and any currently booted instances of that configuration, but IMO it makes more sense to keep them separate.

**Reply by daavoo** [reply]
I also think that makes sense to keep `list`/`status` separated.

A little bit off-topic: we currently have `dvc remote list` but we check the remote status with `dvc status --remote`. 

Would it make sense to have `dvc status --machine` instead of `dvc machine status`? 
If not, would it make sense to move `dvc status --remote` to separate command (`dvc remote status`)?

We have plenty of these ambiguities between semantically similar commands that are exposed differently.

**Reply by pmrowla** [reply]
@daavoo I think we have too many different behaviors within `dvc status` already. We've already had some discussion about revisiting how `dvc status` works (particularly with regard to separating out "workspace vs local cache status" and "local cache vs remote(s) status").

So yes, to answer your question, it may make more sense for us to have `dvc remote status` instead of `status -c --remote` in the future.

**Reply by dberenbaum** [reply]
Good points about remote status, but let's discuss outside of this thread 🙏 .

> * A common configuration for a remote `machine` type (i.e. cloud, CPU size/count, gpu size/type, storage size, etc). This is currently configured via the `config`/`remote` equivalent commands - `dvc machine list/add/remove`
> 
> * One or more started/built/created instances of a `machine` configuration. This is currently configured via the `machine create/destroy/status` commands.

I agree with @jorgeorpinel that this seems clear only from the developer perspective. `add/create`, `remove/destroy`, and `list/status` are each similar. How is a user supposed to know/remember which of these commands are about the config and which are about instances? It would be more clear if we had separate command namespaces, like `dvc machine resource/config list/add/remove` and `dvc machine instance create/destroy/status`. Is that too long? Maybe better naming like `dvc machine build` helps intuitively suggest what each does?

**Reply by karajan1001** [reply]
I agreed to keep `list/status` separated.

But for `status` another choice is to be called `ps`, this name is wildly used in `Linux`, `Docker`, and `Heroku`. The only problem here is that `ps` means `processes` while in `dvc machine` we call it executors.

**Reply by dberenbaum** [reply]
Docker added `docker container list/ls` to replace `docker ps` for basically the same reasons we are discussing: https://stackoverflow.com/questions/45254677/is-there-a-difference-between-docker-ps-and-docker-container-ls. So what do you think about separating into two parts like `dvc machine config/instance` or `dvc machine/instance`?

**Reply by karajan1001** [reply]
But `docker container list/ls` is on level 2 while `dvc machine config/instance ls` is on level three. Maybe we can decouple the remote execution with DVC and build a new project as I don't think remote execution has any strong correlation to data management or maybe we can simply use a `dvc-machine` cmd to reduce the levels.

**Reply by dberenbaum** [reply]
We could make remote execution a separate component, but I think the commands will continue to be under `dvc`. DVC is already about a lot of things outside of data management, and the name is outdated. It really covers anything related to model development/training, including data management, pipelines, experiments, and remote execution. 

What do you think about `dvc machine/instance` since that would be on level 2? Seems similar to docker image/container.

**Reply by casperdcl** [reply]
PS there's also `docker image ls` vs `docker images` :)

**Reply by karajan1001** [reply]
> What do you think about `dvc machine/instance` since that would be on level 2? Seems similar to docker image/container.

`dvc instance` may be more confusing to me, it looks like the instance is a `DVC` instance not a machine instance.

**Reply by dberenbaum** [reply]
A DVC instance would make me think of a compute server to run DVC jobs, which is a pretty accurate description. I'd be more confused by `dvc machine` because it sounds so similar to `dvc instance`.  Maybe we do need to revisit the `dvc machine` naming since it's so vague, but I'm not sure what a better name would be.

**Reply by pmrowla** [reply]
We could go with `dvc resource` and `dvc instance` to match the TPI naming, but I'm not sure that `resource` would be intuitive here unless you were already aware it was specifically tied to the terraform backend

**Reply by dberenbaum** [reply]
I vote to add the extra level but keep the names short, like `dvc mach res/conf/spec/infra/type add` and `dvc mach inst build`.

The `dvc-machine` idea is interesting, but I think it creates confusion for users, like whether they need to install it separately, and it reduces visibility since they already need to be aware of the command to find it.

**Reply by karajan1001** [reply]
How about `dvc mach-inst/mach-res`

**Reply by dberenbaum** [reply]
I don't have a strong preference between `dvc mach inst` and `dvc mach-inst`. Seeing them written out, my concern with both is that it's unclear what they do until you already know what they are short for.

Another option is to make `machine` == instance and organize like:

* dvc machine
  * spec
    * add
    * remove
    * ...
  * build
  * destroy
  * ...

(or have a separate `dvc machine-spec` command)

Some nice things about this layout:
* The spec commands should be less frequently used than the instance commands.
* `instance` is a tricky name -- does it make sense if we support k8s clusters or on-prem hpc in the future?
* `spec` is a commonly understood abbreviation that makes sense.

**Reply by karajan1001** [reply]
> * dvc machine
>   * spec
>   ...

This kind of structure makes sense to me.

**Reply by jorgeorpinel** [reply]
Oops sorry for very late replies... Not sure this is still relevant but re `list/status`:

> we could have a single command that shows both ... but IMO it makes more sense to keep them separate.
I also think that makes sense to keep list/status separated.
I agreed to keep list/status separated.

@pmrowla @daavoo @karajan1001 why do you think so though? I'd err on the side of consolidating unless there's a good reason not to. Also this idea of machines having a status is misleading as they're assumed to be always on I think. We mean whether an experiment is running on them right? That's narrow enough to include in `machine list` I think.

> we check the remote status with dvc status --remote.

Hmmm AFAIK that shows the project status (compared to remotes), not status of remotes per se (do they even have statuses?) so it seems OK @daavoo.

**Reply by jorgeorpinel** [reply]
> It would be more clear if we had separate command namespaces, like `machine config list/add/remove` and `machine instance create/destroy`. Is that too long?

Too long indeed @dberenbaum. I think a single 2nd level under `dvc machine` can include all the necessary commands in 3 groups: `add/modify/remove` for config, `list` show configs, their instances & status, and `build/destroy` to manage instances. That's all we need right? If so I sustain my [previous vote](https://github.com/iterative/dvc/discussions/6890#discussioncomment-1634023).

> * dvc machine
>  * spec
>    * add
>    * ...
>  * build
>  * ...

Similar to this, but the `spec` level is just skipped for simplicity.

> I vote to add the extra level but keep the names short, like `mach res/conf/spec/infra/type` add and `mach inst build`.

I don't mind the `mach` name (sounds like moving very fast, could be a cool feature name) but I'd def. spell out the subcommands ("mach inst" is too cryptic).

> what do you think about separating into two parts ... `machine/instance` ?
> We could go with `resource` and `instance` to match the TPI naming

@dberenbaum @pmrowla `dvc resource` sounds like a more general concept. Maybe remotes are resources, etc. Unless we plan to encapsulate all resources for other features here, I'd try keep all of this under a single level-1 command for now. `dvc machine` or `dvc instance` can both work IMO (one or the other).

---

Summarizing my proposal:

`dvc mach(ine) add/modify/remove` - config commands (consistent with existing features)
`dvc mach(ine) build/destroy` - manage machines (config 1 - * instance)
`dvc mach(ine) list` - list configs and their instances +"status" (whether exps are running somewhere)

**Comment by jorgeorpinel** [comment]
> confusing in the current unreleased documents

Meaning https://www.notion.so/iterative/215a0767aa0341d892b52767e58e2ac6 (outdated) and https://github.com/iterative/dvc/wiki/Remote-executors (incomplete) ? Or are there any more internal docs somewhere

There's also https://github.com/iterative/terraform-provider-iterative/blob/master/README.md which isn't internal and has the same issue to some extent (especially with term "machine", since "instance" has a clear meaning within that context).

**Reply by jorgeorpinel** [reply]
But for the latter (TPI) it's not super clear what the difference between "runner" and "machine" is, other than one term being used for CML (cc @casperdcl @DavidGOrtega) and another one for DVC.

**Reply by DavidGOrtega** [reply]
in TPI `machine` is an instance in the cloud while `runner` is a SCM runner launched inside a `machine`

**Reply by casperdcl** [reply]
CML/TPI uses:

- runner
  + `cml` subcommand name
  + CI provider concept (GH/GL/BB)
- executor
  + CI provider backend orchestrator (mostly just GL)
- resource
  + config entry (within `main.tf` files)
- instance
  + compute provider server (eg AWS EC2 `t2.micro`)
- machine
  + not really used afaik

**Reply by jorgeorpinel** [reply]
> runner is a SCM runner

What is an SCM runner @DavidGOrtega ? Mostly curious. Thanks

**Reply by jorgeorpinel** [reply]
> machine
not really used afaik

@casperdcl yeah I only see the term in the notes for `instance_type`: "You can also specify vendor specific machines in AWS"

**Reply by casperdcl** [reply]
hmm yes we should be more careful to avoid unneeded synonyms...

**Reply by pmrowla** [reply]
> What is an SCM runner

@jorgeorpinel "runner" is what's used by github/gitlab for the machine/environment that runs a github actions/gitlab CI action (https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners)

**Comment by dberenbaum** [comment]
Here's the related references on subcommands in CLI design: https://clig.dev/#subcommands. Also, off topic, I like this diagram:

https://cloud.ibm.com/docs/containers?topic=containers-kubernetes-service-cli

@jorgeorpinel What do you think about something like that in the docs?
